### PR TITLE
Route Filter APIs

### PR DESCRIPTION
**Pull request by dihan0604**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute.Test/Commands.Compute.Test.csproj [hunk 1]
```diff
@@ -67,8 +67,8 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Compute.14.0.0-prerelease\lib\net45\Microsoft.Azure.Management.Compute.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=8.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.1.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=8.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.2.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute.Test/Commands.Compute.Test.csproj]: @dihan0604 this should be 8.3.0-preview now

File: src/ResourceManager/Compute/Commands.Compute.Test/packages.config [hunk 2]
```diff
@@ -7,7 +7,7 @@
   <package id="Microsoft.Azure.Graph.RBAC" version="3.2.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Authorization" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Compute" version="14.0.0-prerelease" targetFramework="net45" />
-  <package id="Microsoft.Azure.Management.Network" version="8.1.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.Network" version="8.2.0-preview" targetFramework="net45" />
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute.Test/packages.config]: @dihan0604 version bump

File: src/ResourceManager/Compute/Commands.Compute/Commands.Compute.csproj [hunk 3]
```diff
@@ -80,8 +80,8 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Compute.14.0.0-prerelease\lib\net45\Microsoft.Azure.Management.Compute.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=8.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.1.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=8.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.2.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Commands.Compute.csproj]: @dihan0604 version bump

File: src/ResourceManager/Compute/Commands.Compute/packages.config [hunk 4]
```diff
@@ -9,7 +9,7 @@
   <package id="Microsoft.Azure.KeyVault.Core" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Authorization" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Compute" version="14.0.0-prerelease" targetFramework="net45" />
-  <package id="Microsoft.Azure.Management.Network" version="8.1.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.Network" version="8.2.0-preview" targetFramework="net45" />
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/packages.config]: @dihan0604 version bump

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 5]
```diff
@@ -67,8 +67,8 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.Insights.0.13.1-preview\lib\net45\Microsoft.Azure.Insights.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=8.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.1.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=8.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.2.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: @dihan0604 version bump

File: src/ResourceManager/Network/Commands.Network.Test/packages.config [hunk 6]
```diff
@@ -7,7 +7,7 @@
   <package id="Microsoft.Azure.Graph.RBAC" version="3.2.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Insights" version="0.13.1-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Authorization" version="1.0.0" targetFramework="net45" />
-  <package id="Microsoft.Azure.Management.Network" version="8.1.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.Network" version="8.2.0-preview" targetFramework="net45" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/packages.config]: @dihan0604 version bump

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj [hunk 7]
```diff
@@ -73,8 +73,8 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Authorization.2.0.0\lib\net40\Microsoft.Azure.Management.Authorization.dll</HintPath>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=8.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.1.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=8.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.8.2.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: @dihan0604 version bump

File: src/ResourceManager/Network/Commands.Network/packages.config [hunk 8]
```diff
@@ -8,7 +8,7 @@
   <package id="Microsoft.Azure.Graph.RBAC" version="3.2.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.KeyVault.Core" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Authorization" version="1.0.0" targetFramework="net45" />
-  <package id="Microsoft.Azure.Management.Network" version="8.1.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.Network" version="8.2.0-preview" targetFramework="net45" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/packages.config]: @dihan0604 version bump

**Comment by dihan0604** [comment]
@cormacpayne , comments addressed

**Comment by dihan0604** [comment]
Created Issue https://github.com/Azure/azure-powershell/issues/3515 and assign to myself. I will add test coverage once the feature is enabled on prod, which should be next week

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1398/

File: src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/AddAzureRouteFilterRuleConfigCommand.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            Mandatory = true,
+            ValueFromPipeline = true,
+            HelpMessage = "The RouteFilter")]
+        public PSRouteFilter RouteFilter { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Do not ask for confirmation if you want to overrite a resource")]
+        public SwitchParameter Force { get; set; }
+        public override void Execute()
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/AddAzureRouteFilterRuleConfigCommand.cs]: @dihan0604 `ConfirmAction` method needs to wrap the functionality that makes changes in this cmdlet

File: src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/NewAzureRouteFilterRuleConfigCommand.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+    [Cmdlet(VerbsCommon.New, "AzureRmRouteFilterRuleConfig", SupportsShouldProcess = true), OutputType(typeof(PSRouteFilterRule))]
+    public class NewAzureRouteFilterRuleConfigCommand : AzureRouteFilterRuleConfigBase
+    {
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Do not ask for confirmation if you want to overrite a resource")]
+        public SwitchParameter Force { get; set; }
+
+        public override void Execute()
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/NewAzureRouteFilterRuleConfigCommand.cs]: @dihan0604 `ConfirmAction` missing

File: src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/RemoveAzureRouteFilterRuleConfigCommand.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            Mandatory = true,
+            ValueFromPipeline = true,
+            HelpMessage = "The RouteFilter")]
+        public PSRouteFilter RouteFilter { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Do not ask for confirmation if you want to overrite a resource")]
+        public SwitchParameter Force { get; set; }
+        public override void Execute()
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/RemoveAzureRouteFilterRuleConfigCommand.cs]: @dihan0604 `ConfirmAction` missing

File: src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/SetAzureRouteFilterRuleConfigCommand.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            Mandatory = true,
+            ValueFromPipeline = true,
+            HelpMessage = "The RouteFilter")]
+        public PSRouteFilter RouteFilter { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Do not ask for confirmation if you want to overrite a resource")]
+        public SwitchParameter Force { get; set; }
+        public override void Execute()
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/RouteFilter/RouteFilterRule/SetAzureRouteFilterRuleConfigCommand.cs]: @dihan0604 `ConfirmAction` missing

File: src/ResourceManager/Network/Commands.Network/RouteFilter/SetAzureRouteFilterCommand.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    [Cmdlet(VerbsCommon.Set, "AzureRmRouteFilter", SupportsShouldProcess = true), OutputType(typeof(PSRouteFilter))]
+    public class SetAzureRouteFilterCommand : RouteFilterBaseCmdlet
+    {
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipeline = true,
+            HelpMessage = "The RouteFilter")]
+        public PSRouteFilter RouteFilter { get; set; }
+
+        public override void Execute()
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/RouteFilter/SetAzureRouteFilterCommand.cs]: @dihan0604 `ConfirmAction` missing

**Comment by dihan0604** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1404/

**Comment by dihan0604** [comment]
On demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1405/

**Comment by dihan0604** [comment]
On demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1411/

**Pull request merged**